Docstrings for proc macros compile error when > 255 characters

I saw the the proc macro support for docstrings was merged recently so I decided to test it in some projects, and I noticed that any  docstring longer than 255 characters will fail to compile with the following error:

```
error[E0080]: evaluation of constant value failed
   --> /Users/dani/Desktop/uniffi-rs/uniffi_core/src/metadata.rs:178:9
    |
178 |         assert!(string.len() < 256);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: string.len() < 256'
    |
note: inside `MetadataBuffer::concat_str`
```

This is the code used to test this:
```rust
/// This is a very long multi line test docstring that exceeds 255 characters.
/// This is a very long multi line test docstring that exceeds 255 characters.
/// This is a very long multi line test docstring that exceeds 255 characters.
/// This is a very long multi line test docstring that exceeds 255 characters.
#[uniffi::export]
pub fn test2() {
}
```

This seems to be caused by the way `MetadataBuffer` stores strings, by encoding their length as a `u8`. This limits their max length to 255, which is easily exceeded if the docstring contains some complex explanation or code example.

My first thought would be to update `MetadataBuffer::concat_str` to encode the length as a `u16` instead, and change `MetadataReader` to do the same, but I'm not sure if that will have some adverse effects somewhere else.

--- a/fixtures/docstring-proc-macro/src/lib.rs
+++ b/fixtures/docstring-proc-macro/src/lib.rs
@@ -94,4 +94,11 @@ pub trait CallbackTest {
     fn test(&self);
 }
 
+/// This is a very long multi line test docstring that exceeds 255 characters.
+/// This is a very long multi line test docstring that exceeds 255 characters.
+/// This is a very long multi line test docstring that exceeds 255 characters.
+/// This is a very long multi line test docstring that exceeds 255 characters.
+#[uniffi::export]
+pub fn test_long_docstring() {}
+
 uniffi::include_scaffolding!("docstring-proc-macro");

--- a/uniffi_core/src/metadata.rs
+++ b/uniffi_core/src/metadata.rs
@@ -168,7 +168,8 @@ impl MetadataBuffer {
         self.concat_value(value as u8)
     }
 
-    // Concatenate a string to this buffer.
+    // Concatenate a string to this buffer. The maximum string length is 255 bytes. For longer strings,
+    // use `concat_long_str()`.
     //
     // Strings are encoded as a `u8` length, followed by the utf8 data.
     //
@@ -179,6 +180,28 @@ impl MetadataBuffer {
         assert!(self.size + string.len() < BUF_SIZE);
         self.bytes[self.size] = string.len() as u8;
         self.size += 1;
+        let bytes = string.as_bytes();
+        let mut i = 0;
+        while i < bytes.len() {
+            self.bytes[self.size] = bytes[i];
+            self.size += 1;
+            i += 1;
+        }
+        self
+    }
+
+    // Concatenate a longer string to this buffer.
+    //
+    // Strings are encoded as a `u16` length, followed by the utf8 data.
+    //
+    // This consumes self, which is convenient for the proc-macro code and also allows us to avoid
+    // allocated an extra buffer.
+    pub const fn concat_long_str(mut self, string: &str) -> Self {
+        assert!(self.size + string.len() + 1 < BUF_SIZE);
+        let [lo, hi] = (string.len() as u16).to_le_bytes();
+        self.bytes[self.size] = lo;
+        self.bytes[self.size + 1] = hi;
+        self.size += 2;
         let bytes = string.as_bytes();
         let mut i = 0;
         while i < bytes.len() {

--- a/uniffi_macros/src/enum_.rs
+++ b/uniffi_macros/src/enum_.rs
@@ -152,7 +152,7 @@ pub(crate) fn enum_meta_static_var(
             .concat_str(#name)
     };
     metadata_expr.extend(variant_metadata(enum_)?);
-    metadata_expr.extend(quote! { .concat_str(#docstring) });
+    metadata_expr.extend(quote! { .concat_long_str(#docstring) });
     Ok(create_metadata_items("enum", &name, metadata_expr, None))
 }
 
@@ -200,9 +200,9 @@ pub fn variant_metadata(enum_: &DataEnum) -> syn::Result<Vec<TokenStream>> {
                                 .concat(<#field_types as ::uniffi::Lower<crate::UniFfiTag>>::TYPE_ID_META)
                                 // field defaults not yet supported for enums
                                 .concat_bool(false)
-                                .concat_str(#field_docstrings)
+                                .concat_long_str(#field_docstrings)
                             )*
-                        .concat_str(#docstring)
+                        .concat_long_str(#docstring)
                     })
                 })
         )

--- a/uniffi_macros/src/error.rs
+++ b/uniffi_macros/src/error.rs
@@ -212,7 +212,7 @@ pub(crate) fn error_meta_static_var(
     } else {
         metadata_expr.extend(variant_metadata(enum_)?);
     }
-    metadata_expr.extend(quote! { .concat_str(#docstring) });
+    metadata_expr.extend(quote! { .concat_long_str(#docstring) });
     Ok(create_metadata_items("error", &name, metadata_expr, None))
 }
 
@@ -225,7 +225,7 @@ pub fn flat_error_variant_metadata(enum_: &DataEnum) -> syn::Result<Vec<TokenStr
             let docstring = extract_docstring(&v.attrs)?;
             Ok(quote! {
                 .concat_str(#name)
-                .concat_str(#docstring)
+                .concat_long_str(#docstring)
             })
         }))
         .collect()

--- a/uniffi_macros/src/export/callback_interface.rs
+++ b/uniffi_macros/src/export/callback_interface.rs
@@ -192,7 +192,7 @@ pub(super) fn metadata_items(
             ::uniffi::MetadataBuffer::from_code(::uniffi::metadata::codes::CALLBACK_INTERFACE)
                 .concat_str(#module_path)
                 .concat_str(#trait_name)
-                .concat_str(#docstring)
+                .concat_long_str(#docstring)
         },
         None,
     );

--- a/uniffi_macros/src/fnsig.rs
+++ b/uniffi_macros/src/fnsig.rs
@@ -224,7 +224,7 @@ impl FnSignature {
                     .concat_value(#args_len)
                     #(#arg_metadata_calls)*
                     .concat(<#return_ty as ::uniffi::LowerReturn<crate::UniFfiTag>>::TYPE_ID_META)
-                    .concat_str(#docstring)
+                    .concat_long_str(#docstring)
             }),
 
             FnKind::Method { self_ident } => {
@@ -238,7 +238,7 @@ impl FnSignature {
                         .concat_value(#args_len)
                         #(#arg_metadata_calls)*
                         .concat(<#return_ty as ::uniffi::LowerReturn<crate::UniFfiTag>>::TYPE_ID_META)
-                        .concat_str(#docstring)
+                        .concat_long_str(#docstring)
                 })
             }
 
@@ -254,7 +254,7 @@ impl FnSignature {
                         .concat_value(#args_len)
                         #(#arg_metadata_calls)*
                         .concat(<#return_ty as ::uniffi::LowerReturn<crate::UniFfiTag>>::TYPE_ID_META)
-                        .concat_str(#docstring)
+                        .concat_long_str(#docstring)
                 })
             }
 
@@ -268,7 +268,7 @@ impl FnSignature {
                         .concat_value(#args_len)
                         #(#arg_metadata_calls)*
                         .concat(<#return_ty as ::uniffi::LowerReturn<crate::UniFfiTag>>::TYPE_ID_META)
-                        .concat_str(#docstring)
+                        .concat_long_str(#docstring)
                 })
             }
         }

--- a/uniffi_macros/src/object.rs
+++ b/uniffi_macros/src/object.rs
@@ -146,7 +146,7 @@ pub(crate) fn interface_meta_static_var(
                 .concat_str(#module_path)
                 .concat_str(#name)
                 .concat_bool(#is_trait)
-                .concat_str(#docstring)
+                .concat_long_str(#docstring)
         },
         None,
     ))

--- a/uniffi_macros/src/record.rs
+++ b/uniffi_macros/src/record.rs
@@ -166,7 +166,7 @@ pub(crate) fn record_meta_static_var(
                 .concat_str(#name)
                 .concat(<#ty as ::uniffi::Lower<crate::UniFfiTag>>::TYPE_ID_META)
                 #default
-                .concat_str(#docstring)
+                .concat_long_str(#docstring)
             })
         })
         .collect::<syn::Result<_>>()?;
@@ -180,7 +180,7 @@ pub(crate) fn record_meta_static_var(
                 .concat_str(#name)
                 .concat_value(#fields_len)
                 #concat_fields
-                .concat_str(#docstring)
+                .concat_long_str(#docstring)
         },
         None,
     ))

--- a/uniffi_meta/src/reader.rs
+++ b/uniffi_meta/src/reader.rs
@@ -80,6 +80,17 @@ impl<'a> MetadataReader<'a> {
         }
     }
 
+    fn read_u16(&mut self) -> Result<u16> {
+        if self.buf.len() >= 2 {
+            // read the value as little-endian
+            let value = u16::from_le_bytes([self.buf[0], self.buf[1]]);
+            self.buf = &self.buf[2..];
+            Ok(value)
+        } else {
+            bail!("Not enough data left in buffer to read a u16 value");
+        }
+    }
+
     fn read_u32(&mut self) -> Result<u32> {
         if self.buf.len() >= 4 {
             // read the value as little-endian
@@ -105,8 +116,15 @@ impl<'a> MetadataReader<'a> {
         String::from_utf8(slice.into()).context("Invalid string data")
     }
 
-    fn read_optional_string(&mut self) -> Result<Option<String>> {
-        Ok(Some(self.read_string()?).filter(|str| !str.is_empty()))
+    fn read_long_string(&mut self) -> Result<String> {
+        let size = self.read_u16()? as usize;
+        let slice;
+        (slice, self.buf) = self.buf.split_at(size);
+        String::from_utf8(slice.into()).context("Invalid string data")
+    }
+
+    fn read_optional_long_string(&mut self) -> Result<Option<String>> {
+        Ok(Some(self.read_long_string()?).filter(|str| !str.is_empty()))
     }
 
     fn read_type(&mut self) -> Result<Type> {
@@ -202,7 +220,7 @@ impl<'a> MetadataReader<'a> {
         let is_async = self.read_bool()?;
         let inputs = self.read_inputs()?;
         let (return_type, throws) = self.read_return_type()?;
-        let docstring = self.read_optional_string()?;
+        let docstring = self.read_optional_long_string()?;
         Ok(FnMetadata {
             module_path,
             name,
@@ -221,7 +239,7 @@ impl<'a> MetadataReader<'a> {
         let name = self.read_string()?;
         let inputs = self.read_inputs()?;
         let (return_type, throws) = self.read_return_type()?;
-        let docstring = self.read_optional_string()?;
+        let docstring = self.read_optional_long_string()?;
 
         return_type
             .filter(|t| {
@@ -250,7 +268,7 @@ impl<'a> MetadataReader<'a> {
         let is_async = self.read_bool()?;
         let inputs = self.read_inputs()?;
         let (return_type, throws) = self.read_return_type()?;
-        let docstring = self.read_optional_string()?;
+        let docstring = self.read_optional_long_string()?;
         Ok(MethodMetadata {
             module_path,
             self_name,
@@ -270,7 +288,7 @@ impl<'a> MetadataReader<'a> {
             module_path: self.read_string()?,
             name: self.read_string()?,
             fields: self.read_fields()?,
-            docstring: self.read_optional_string()?,
+            docstring: self.read_optional_long_string()?,
         })
     }
 
@@ -287,7 +305,7 @@ impl<'a> MetadataReader<'a> {
             module_path,
             name,
             variants,
-            docstring: self.read_optional_string()?,
+            docstring: self.read_optional_long_string()?,
         })
     }
 
@@ -302,7 +320,7 @@ impl<'a> MetadataReader<'a> {
             module_path: self.read_string()?,
             name: self.read_string()?,
             imp: ObjectImpl::from_is_trait(self.read_bool()?),
-            docstring: self.read_optional_string()?,
+            docstring: self.read_optional_long_string()?,
         })
     }
 
@@ -335,7 +353,7 @@ impl<'a> MetadataReader<'a> {
         Ok(CallbackInterfaceMetadata {
             module_path: self.read_string()?,
             name: self.read_string()?,
-            docstring: self.read_optional_string()?,
+            docstring: self.read_optional_long_string()?,
         })
     }
 
@@ -347,7 +365,7 @@ impl<'a> MetadataReader<'a> {
         let is_async = self.read_bool()?;
         let inputs = self.read_inputs()?;
         let (return_type, throws) = self.read_return_type()?;
-        let docstring = self.read_optional_string()?;
+        let docstring = self.read_optional_long_string()?;
         Ok(TraitMethodMetadata {
             module_path,
             trait_name,
@@ -374,7 +392,7 @@ impl<'a> MetadataReader<'a> {
                     name,
                     ty,
                     default,
-                    docstring: self.read_optional_string()?,
+                    docstring: self.read_optional_long_string()?,
                 })
             })
             .collect()
@@ -387,7 +405,7 @@ impl<'a> MetadataReader<'a> {
                 Ok(VariantMetadata {
                     name: self.read_string()?,
                     fields: self.read_fields()?,
-                    docstring: self.read_optional_string()?,
+                    docstring: self.read_optional_long_string()?,
                 })
             })
             .collect()
@@ -400,7 +418,7 @@ impl<'a> MetadataReader<'a> {
                 Ok(VariantMetadata {
                     name: self.read_string()?,
                     fields: vec![],
-                    docstring: self.read_optional_string()?,
+                    docstring: self.read_optional_long_string()?,
                 })
             })
             .collect()

